### PR TITLE
8260473: [vector] ZGC: VectorReshape test produces incorrect results with ZGC enabled

### PR DESCRIPTION
**Pull request by casparcwang**
https://bugs.openjdk.java.net/browse/JDK-8260473

Function "PhaseVector::expand_vunbox_node" creates a LoadNode, but forgets to make the LoadNode to pass gc barriers.

Testing: all Vector API related tests have passed.

Original pr: https://github.com/openjdk/jdk/pull/2253

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260473](https://bugs.openjdk.java.net/browse/JDK-8260473): [vector] ZGC: VectorReshape test produces incorrect results with ZGC enabled

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Contributors
 * Stuart Monteith `<smonteith@openjdk.org>`
 * Wang Chao `<casparcwang@tencent.com>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk16 pull/139/head:pull/139`
`$ git checkout pull/139`

### CODE REVIEW

File: test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
Comment (iwanowww) [inline test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: Why `-XX:CICompilerCount=1` and `-Xmx256m` are needed?

**Review by iwanowww — APPROVED** [review]
Overall, looks good.

File: test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
Comment (DamonFool) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: > Why `-XX:CICompilerCount=1` and `-Xmx256m` are needed?

Thanks @iwanowww for your review.

I discussed the same question with @casparcwang offline.
The reason is:
 - Small heap (-Xmx256m) will help to trigger a gc.
 - compileonly and compilercount=1 will let the VM run slow enough to wait for a gc to be finished.
 
And @casparcwang told me that this bug seems not to be reproduced every time without these JVM args.
Thanks.

**Review by neliasso — APPROVED** [review]
Approved. 

Now awaiting release team approval.

File: test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
Comment (iwanowww) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: > compileonly and compilercount=1 will let the VM run slow enough to wait for a gc to be finished.

That's a strange way to provoke the bug. You could just increase the number of iterations instead.

But the right way to fix it is to stress ZGC to continuously run in the background while the test case aggressively unboxes vectors in compiled code. `-Xmx256m` helps with that while `-XX:CICompilerCount=1` is irrelevant.

Comment (casparcwang) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: > > compileonly and compilercount=1 will let the VM run slow enough to wait for a gc to be finished.
> 
> That's a strange way to provoke the bug. You could just increase the number of iterations instead.
> 
> But the right way to fix it is to stress ZGC to continuously run in the background while the test case aggressively unboxes vectors in compiled code. `-Xmx256m` helps with that while `-XX:CICompilerCount=1` is irrelevant.

Yes, it's very weird to provoke the bug like this. If CICompilerCount=1 is removed, the test failed 60% roughly on my machine.
And the iteration has already changed from 100 to 1000,  the run time of the test is nearly 30s on release version of jvm.

If I add the following patch, the test always fails on my machine,

```
diff --git a/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java b/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
index 1843ec0..959b29a 100644
--- a/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
+++ b/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
@@ -44,7 +44,7 @@ import jdk.internal.vm.annotation.ForceInline;
  * @modules jdk.incubator.vector
  * @modules java.base/jdk.internal.vm.annotation
  * @run testng/othervm -XX:CompileCommand=compileonly,jdk/incubator/vector/ByteVector.fromByteBuffer
- *      -XX:-TieredCompilation -XX:CICompilerCount=1 -XX:+UseZGC -Xbatch -Xmx256m VectorRebracket128Test
+ *      -XX:-TieredCompilation -XX:+UseZGC -Xmx256m VectorRebracket128Test
  */
 
 @Test
@@ -125,6 +125,14 @@ public class VectorRebracket128Test {
     @ForceInline
     static <E,F>
     void testVectorRebracket(VectorSpecies<E> a, VectorSpecies<F> b, byte[] input, byte[] output) {
+        new Thread(() -> {
+            while (true) {
+                try {
+                    System.gc();
+                    Thread.sleep(100);
+                } catch (Exception e) {}
+            }
+        }).start();
         Vector<E> av = a.fromByteArray(input, 0, ByteOrder.nativeOrder());
         int block;
         assert(input.length == output.length);
```

Comment (casparcwang) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: sorry for the wrong patch above,  the failed reason of the patch above is due to stack creation failure (create 1000 threads). The following is the right stress gc patch.

```
diff --git a/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java b/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
index 6b266db..a761ea2 100644
--- a/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
+++ b/test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
@@ -44,7 +44,7 @@ import jdk.internal.vm.annotation.ForceInline;
  * @modules jdk.incubator.vector
  * @modules java.base/jdk.internal.vm.annotation
  * @run testng/othervm -XX:CompileCommand=compileonly,jdk/incubator/vector/ByteVector.fromByteBuffer
- *      -XX:-TieredCompilation -XX:CICompilerCount=1 -XX:+UseZGC -Xbatch -Xmx256m VectorRebracket128Test
+ *      -XX:-TieredCompilation -XX:+UseZGC -Xmx256m VectorRebracket128Test
  */
 
 @Test
@@ -59,6 +59,19 @@ public class VectorRebracket128Test {
     static final VectorSpecies<Byte> bspec128 = ByteVector.SPECIES_128;
     static final VectorSpecies<Short> sspec128 = ShortVector.SPECIES_128;
 
+    static {
+        Thread t = new Thread(() -> {
+            while (true) {
+                try {
+                    System.gc();
+                    Thread.sleep(100);
+                } catch (Exception e) {}
+            }
+        });
+        t.setDaemon(true);
+        t.start();
+    }
+
     static <T> IntFunction<T> withToString(String s, IntFunction<T> f) {
         return new IntFunction<T>() {
             @Override
```

Comment (iwanowww) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: Good. Please, file a follow-up RFE to improve the test.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: > Good. Please, file a follow-up RFE to improve the test.

OK.
I will help to file a JBS bug once the fix has been merged into the jdk mainline.
It will be only fixed in the jdk17, right?
Thanks.

**Comment by iwanowww** [comment]
> It will be only fixed in the jdk17, right?

Yes, I'm OK with that.

File: test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java
Comment (DamonFool) [reply test/hotspot/jtreg/compiler/vectorapi/VectorRebracket128Test.java:47]: > Good. Please, file a follow-up RFE to improve the test.

The RFE has been filed here: https://bugs.openjdk.java.net/browse/JDK-8261152
Thanks.

**Pull request closed without merging**
